### PR TITLE
feat(explain): reimplement explain using pretty-xmlish

### PR DESCRIPTION
**Pull request by ice1000**
Edit: here's an example from planner tests:

```
Projection { cost: 8163.5923, exprs: [ sum {  } ] }                                                                 
└── Aggregate { aggs: [ sum {  } ], cost: 8161.992, group_by: [] }                                                  
    └── Projection { cost: 7964.3843, exprs: [ l_extendedprice, l_discount ] }                                      
        └── Filter                                                                                                  
            ├── cond: and                                                                                           
            │   ├── lhs: >= { lhs: l_shipdate, rhs: 1994-01-01 }                                                    
            │   └── rhs: and                                                                                        
            │       ├── lhs: > { lhs: 1995-01-01, rhs: l_shipdate }                                                 
            │       └── rhs: and                                                                                    
            │           ├── lhs: >= { lhs: 0.09, rhs: l_discount }                                                  
            │           └── rhs: and { lhs: > { lhs: 24, rhs: l_quantity }, rhs: >= { lhs: l_discount, rhs: 0.07 } }
            ├── cost: 7210.72                                                                                       
            └── Scan { cost: 4000, list: [ l_quantity, l_extendedprice, l_discount, l_shipdate ], table: lineitem }
```

Demo:

```
> explain create table t(v1 int not null, v2 int not null);
+---------------------------------------------------------------------------------+
|                                                                                 |
| CreateTable                                                                     |
| ├── columns: [ Column { name: v1, type: INT }, Column { name: v2, type: INT } ] |
| ├── cost: 0.1                                                                   |
| ├── database_id: 0                                                              |
| ├── name: t                                                                     |
| ├── ordered_ids: []                                                             |
| └── schema_id: 0                                                                |
|                                                                                 |
+---------------------------------------------------------------------------------+
in 0.025s
> create table t(v1 int not null, v2 int not null);
+---+
| 1 |
+---+
in 0.003s
> explain insert into t values(1, 1), (4, 2), (3, 3), (10, 12), (2, 5);
+-------------------------------------------------+
|                                                 |
| Insert { cols: [ v1, v2 ], cost: 20, table: t } |
| └── Values { cost: 10, rows: 5 }                |
|     ├── [ 1, 1 ]                                |
|     ├── [ 4, 2 ]                                |
|     ├── [ 3, 3 ]                                |
|     ├── [ 10, 12 ]                              |
|     └── [ 2, 5 ]                                |
|                                                 |
+-------------------------------------------------+
in 0.006s
> create table ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt(v1 int not null, v2 int not null);
+---+
| 1 |
+---+
in 0.003s
> explain insert into ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt values(1, 1), (4, 2), (3, 3), (10, 12), (2, 5);
+--------------------------------------------------------------------------------------------+
|                                                                                            |
| Insert                                                                                     |
| ├── cols: [ v1, v2 ]                                                                       |
| ├── cost: 20                                                                               |
| ├── table: ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt |
| └── Values { cost: 10, rows: 5 }                                                           |
|     ├── [ 1, 1 ]                                                                           |
|     ├── [ 4, 2 ]                                                                           |
|     ├── [ 3, 3 ]                                                                           |
|     ├── [ 10, 12 ]                                                                         |
|     └── [ 2, 5 ]                                                                           |
|                                                                                            |
+--------------------------------------------------------------------------------------------+
in 0.006s
```

Before this PR, we have:

```
> explain create table t(v1 int not null, v2 int not null);
+---------------------------------------+
| CreateTable: name="t", ... (cost=0.1) |
|                                       |
+---------------------------------------+
in 0.024s
> create table t(v1 int not null, v2 int not null);
+---+
| 1 |
+---+
in 0.003s
> explain insert into t values(1, 1), (4, 2), (3, 3), (10, 12), (2, 5);
+-----------------------------+
| Insert: t[v1, v2] (cost=20) |
|   Values: 5 rows (cost=10)  |
|                             |
+-----------------------------+
in 0.006s
> create table ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt(v1 int not null, v2 int not null);
+---+
| 1 |
+---+
in 0.002s
> explain insert into ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt values(1, 1), (4, 2), (3, 3), (10, 12), (2, 5);
+-----------------------------------------------------------------------------------------------------------+
| Insert: ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt[v1, v2] (cost=20) |
|   Values: 5 rows (cost=10)                                                                                |
|                                                                                                           |
+-----------------------------------------------------------------------------------------------------------+
in 0.006s
```

This is a first demonstration of pretty-xmlish. There are some improvements I plan to do with the library so there will be another PR after this.

### CODE REVIEW

**Comment by ice1000** [comment]
Again need help on fixing tests...

**Comment by xxchan** [comment]
What help do you need? 👀

File: src/planner/explain.rs [hunk 1]
```diff
@@ -65,210 +83,261 @@ impl<'a> Explain<'a> {
...
+    fn cost(&self) -> Option<f32> {
+        self.costs.map(|cs| cs[usize::from(self.id)])
     }
 
     /// Returns whether the expression is `true`.
     #[inline]
     fn is_true(&self, id: &Id) -> bool {
         self.expr[*id] == Expr::true_()
     }
-}
```
[the remark this answers is not in the document]
Comment (xxchan) [reply src/planner/explain.rs:100]:
```suggestion
    /// Transforms the plan to `Pretty`, an intermediate representation for pretty printing.
```

People may be unfamiliar with the concepts of pretty printing, so it might be better to add some doc comments

File: src/utils/mod.rs [hunk 2]
```diff
@@ -1,3 +1,19 @@
...
+
+pub mod pretty {
+    use std::borrow::Cow;
+    use std::collections::BTreeMap;
+
+    use pretty_xmlish::{Pretty, XmlNode};
+
+    pub fn named_record<'a>(
+        name: impl Into<Cow<'a, str>>,
+        fields: BTreeMap<&'a str, Pretty<'a>>,
```
Comment (xxchan) [inline src/utils/mod.rs:13]: Should we use a map preserving insertion order instead? 

Currently it is printed in dict order.

Comment (xxchan) [reply src/utils/mod.rs:13]: Or maybe simply `Vec<(_,_)>` ...

**Comment by ice1000** [comment]
> What help do you need? 👀

I expect the test data to be somehow generated, but I don't know how 😳

Sorry, I should've explained earlier.

**Comment by xxchan** [comment]
> I expect the test data to be somehow generated, but I don't know how 😳

Do you want `make apply_planner_test`?

File: src/binder_v2/create_table.rs [hunk 3]
```diff
@@ -16,13 +20,29 @@ pub struct CreateTable {
...
+        let config = PrettyConfig {
+            need_boundaries: false,
+            ..PrettyConfig::default()
+        };
+        config.unicode(&mut explain, &explainer);
+        f.write_str(&explain)
+    }
+}
+
+// TODO: rewrite display with pretty_xmlish
```
[the remark this answers is not in the document]
Comment (ice1000) [reply src/binder_v2/create_table.rs:36]: Nope... It requires updating the crate itself.

[the remark this answers is not in the document]
Comment (ice1000) [reply src/binder_v2/create_table.rs:36]: I expected a one-linear in the impl of `fmt`, so there needs to be a helper function in the crate

**Comment by xxchan** [comment]
Others generally feel good to me! As a person not very familiar with pretty-printing, the major benefits I felt are:
- Adding one layer of IR separates concern of "what to print" and "how to print", and makes implementation easier. 
- The wrapping behavior is cool: (it's worth adding to the doc for `pretty-xmlish` 🤔)
```
+--------------------------------------------------------------------------------------------+
|                                                                                            |
| Insert                                                                                     |
| ├── cols: [ v1, v2 ]                                                                       |
| ├── cost: 20                                                                               |
| ├── table: ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt |
| └── Values { cost: 10, rows: 5 }                                                           |
|     ├── [ 1, 1 ]                                                                           |
|     ├── [ 4, 2 ]                                                                           |
|     ├── [ 3, 3 ]                                                                           |
|     ├── [ 10, 12 ]                                                                         |
|     └── [ 2, 5 ]                                                                           |
|                                                                                            |
+--------------------------------------------------------------------------------------------+
```

BTW, I still don't fully understand what's the differences between `pretty-xmlish` and `pretty`, and other benefits of `pretty-xmlish`. To gain more adoption, it might be better to make it more clear and include some examples in the future. 🫣

**Comment by ice1000** [comment]
> it's worth adding to the doc for `pretty-xmlish` 🤔

What do you mean 'doc'? I have an RFC for `pretty-xmlish` but idk why it's not on crates.io...

**Comment by ice1000** [comment]
Short answer: when you need the `└──` thingy, it requires a stack in the textualizer, which is impossible with `pretty`.

File: src/catalog/column.rs [hunk 4]
```diff
@@ -81,6 +98,10 @@ impl ColumnCatalog {
         &self.desc.name
     }
 
+    pub(crate) fn into_name(self) -> String {
```
[the remark this answers is not in the document]
Comment (ice1000) [reply src/catalog/column.rs:101]: lmao I agree, but I am concerned about the overall influence. Will address it later

File: src/utils/mod.rs (hunk 2 above)
Comment (ice1000) [reply src/utils/mod.rs:13]: I think there should be a generic type for records, but Idk about these practices in Rust. In Java there's a subtype called `LinkedHashMap`

File: src/binder_v2/create_table.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ice1000) [reply src/binder_v2/create_table.rs]: Fixed

File: src/planner/explain.rs [hunk 5]
```diff
@@ -65,210 +83,261 @@ impl<'a> Explain<'a> {
...
             | Eq([a, b]) | NotEq([a, b]) | And([a, b]) | Or([a, b]) | Xor([a, b])
-            | Like([a, b]) => write!(f, "({} {} {})", self.expr(a), enode, self.expr(b)),
+            | Like([a, b]) => named_record(
+                enode.to_string(),
+                btreemap! {
+                    "lhs" => self.expr(a).pretty(),
+                    "rhs" => self.expr(b).pretty(),
+                },
+                vec![],
+            ),
```
[the remark this answers is not in the document]
Comment (ice1000) [reply src/planner/explain.rs]: I agree, but it's quite complicated... It would be nice if we can do a precedence-driven pretty printing, but that's another story and requires a major rewrite of the data structure (I believe)

File: src/utils/mod.rs (hunk 2 above)
Comment (ice1000) [reply src/utils/mod.rs]: I think it's fine to use dict order

File: src/planner/explain.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ice1000) [reply src/planner/explain.rs]: My crate simply does not handle these simple cases, because inline expressions should be separated from larger data structures, for instance you don't want `Add(Record(), Scan())` to be a valid expression, and it would be nice if we can rule this monstrosity out by the type system..

File: src/executor_v2/explain.rs [hunk 6]
```diff
@@ -14,12 +15,17 @@ pub struct ExplainExecutor {
...
 impl ExplainExecutor {
     pub fn execute(self) -> BoxedExecutor {
-        let explain = format!(
-            "{}",
-            Explain::of(&self.plan)
-                .with_costs(&costs(&self.plan))
-                .with_catalog(&self.catalog)
-        );
+        let binding = costs(&self.plan);
+        let binding = Explain::of(&self.plan)
```
Comment (wangrunji0408) [inline src/executor_v2/explain.rs:19]: Why called 'binding'?

File: tests/planner_test/tpch.planner.sql [hunk 7]
```diff
@@ -112,12 +112,59 @@ order by
...
-      Projection: [l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus] (cost=18260)
-        Filter: (1998-09-21 >= l_shipdate) (cost=12900)
-          Scan: lineitem[l_quantity, l_extendedprice, l_discount, l_tax, l_returnflag, l_linestatus, l_shipdate] (cost=7000)
+
+Projection                                                                                                              
+├── cost: 39238.984                                                                                                     
+├── exprs:                                                                                                              
+│   ┌── l_returnflag                                                                                                    
+│   ├── l_linestatus                                                                                                    
+│   ├── sum {  }
```
Comment (wangrunji0408) [inline tests/planner_test/tpch.planner.sql:121]: The inner expression is missing. 😇

BTW, this empty brackets seems a little wide. 😇
How about `{}` or `{ }`?

File: tests/planner_test/tpch.planner.sql [hunk 8]
```diff
@@ -112,12 +112,59 @@ order by
...
+│   ├── l_linestatus                                                                                                    
+│   ├── sum {  }                                                                                                        
+│   │   └── l_quantity                                                                                                  
+│   ├── sum {  }                                                                                                        
+│   │   └── l_extendedprice                                                                                             
+│   ├── sum {  }                                                                                                        
+│   │   └── * { lhs: l_extendedprice, rhs: - { lhs: 1, rhs: l_discount } }                                              
+│   ├── sum {  }                                                                                                        
+│   │   └── * { lhs: + { lhs: l_tax, rhs: 1 }, rhs: * { lhs: l_extendedprice, rhs: - { lhs: 1, rhs: l_discount } } }
```
Comment (wangrunji0408) [inline tests/planner_test/tpch.planner.sql:128]: `lhs` and `rhs` looks verbose in expressions. How about this:
```suggestion
│   │   └── *(+(l_tax, 1), *(l_extendedprice, -(1, l_discount)))   
```

File: tests/planner_test/tpch.planner.sql [hunk 9]
```diff
@@ -112,12 +112,59 @@ order by
...
+│   ├── sum {  }                                                                                                        
+│   │   └── * { lhs: + { lhs: l_tax, rhs: 1 }, rhs: * { lhs: l_extendedprice, rhs: - { lhs: 1, rhs: l_discount } } }    
+│   ├── / { lhs: sum {  }, rhs: count {  } }                                                                            
+                                                                                                                        
+│   ├── / { lhs: sum {  }, rhs: count {  } }                                                                            
+                                                                                                                        
+│   ├── / { lhs: sum {  }, rhs: count {  } }                                                                            
+                                                                                                                        
+│   └── rowcount                                                                                                        
+└── Order { by: [ asc {  }, asc {  } ], cost: 33158.984 }
```
Comment (wangrunji0408) [inline tests/planner_test/tpch.planner.sql:136]: The order keys are missing too.

File: tests/planner_test/tpch.planner.sql [hunk 10]
```diff
@@ -146,21 +193,33 @@ order by
...
+└── TopN { cost: 63647.88, limit: 10, offset: 0, order_by: [ desc {  }, asc {  } ] }                                
+    └── Aggregate { aggs: [ sum {  } ], cost: 62224.105, group_by: [ l_orderkey, o_orderdate, o_shippriority ] }    
+        └── Projection                                                                                              
+            ├── cost: 59744.105                                                                                     
+            ├── exprs: [ o_orderdate, o_shippriority, l_orderkey, l_extendedprice, l_discount ]                     
+            └── HashJoin { cost: 55264.105, on: Equality { lhs: [ o_orderkey ], rhs: [ l_orderkey ] }, type: inner }
+                ├── Projection { cost: 24811.05, exprs: [ o_orderkey, o_orderdate, o_shippriority ] }               
+                │   └── HashJoin                                                                                    
+                │       ├── cost: 22091.05                                                                          
+                │       ├── on: Equality { lhs: [ c_custkey ], rhs: [ o_custkey ] }
```
Comment (wangrunji0408) [inline tests/planner_test/tpch.planner.sql:207]: why not?
```suggestion
                │       ├── on: = { lhs: [ c_custkey ], rhs: [ o_custkey ] }                                 
```

File: tests/planner_test/tpch.planner.sql [hunk 11]
```diff
@@ -146,21 +193,33 @@ order by
...
+            ├── exprs: [ o_orderdate, o_shippriority, l_orderkey, l_extendedprice, l_discount ]                     
+            └── HashJoin { cost: 55264.105, on: Equality { lhs: [ o_orderkey ], rhs: [ l_orderkey ] }, type: inner }
+                ├── Projection { cost: 24811.05, exprs: [ o_orderkey, o_orderdate, o_shippriority ] }               
+                │   └── HashJoin                                                                                    
+                │       ├── cost: 22091.05                                                                          
+                │       ├── on: Equality { lhs: [ c_custkey ], rhs: [ o_custkey ] }                                 
+                │       ├── type: inner                                                                             
+                │       ├── Projection { cost: 2940, exprs: [ c_custkey ] }                                         
+                │       │   └── Filter { cond: = { lhs: c_mktsegment, rhs: 'BUILDING' }, cost: 2700 }               
+                │       │       └── Scan { cost: 2000, list: [ c_custkey, c_mktsegment ], table: customer }
```
Comment (wangrunji0408) [inline tests/planner_test/tpch.planner.sql:211]: Is it possible to keep the cost always at the end?

**Comment by wangrunji0408** [comment]
> > it's worth adding to the doc for `pretty-xmlish` 🤔
> 
> What do you mean 'doc'? I have an RFC for `pretty-xmlish` but idk why it's not on crates.io...

+1 request for the code docs that can be shown in [docs.rs](https://docs.rs/pretty-xmlish/latest/pretty_xmlish/index.html). Now it looks blank. 🫥

**Review by wangrunji0408 — COMMENTED** [review]
The new format looks pretty neat! 😊

File: tests/planner_test/tpch.planner.sql (hunk 8 above)
Comment (ice1000) [reply tests/planner_test/tpch.planner.sql:128]: It is very hard to make a special case for binary expressions. You'll have to change the generator to work for binary expressions in general.

File: tests/planner_test/tpch.planner.sql (hunk 11 above)
Comment (ice1000) [reply tests/planner_test/tpch.planner.sql:211]: Frankly, I don't know a good way. Maybe I should change the data structure from BTreeMap to associate lists ..

File: tests/planner_test/tpch.planner.sql (hunk 10 above)
Comment (ice1000) [reply tests/planner_test/tpch.planner.sql:207]: Could you please propose the change to the generator? I can commit directly with your name if you do so

File: src/executor_v2/explain.rs (hunk 6 above)
Comment (ice1000) [reply src/executor_v2/explain.rs:19]: I think it's not intended. I'll rename them. They're generated by rust analyzer

File: tests/planner_test/tpch.planner.sql (hunk 7 above)
Comment (ice1000) [reply tests/planner_test/tpch.planner.sql:121]: I would say it's probably a bug in the generator. I'll take a look

**Comment by ice1000** [comment]
> > > it's worth adding to the doc for `pretty-xmlish` 🤔
> > 
> > What do you mean 'doc'? I have an RFC for `pretty-xmlish` but idk why it's not on crates.io...
> 
> +1 request for the code docs that can be shown in [docs.rs](https://docs.rs/pretty-xmlish/latest/pretty_xmlish/index.html). Now it looks blank. 🫥

I think the rfc is pretty descriptive. I think it's better than comment-based docs... But it's just my personal opinion.

File: tests/planner_test/tpch.planner.sql (hunk 8 above)
Comment (wangrunji0408) [reply tests/planner_test/tpch.planner.sql:128]: A feature request for pretty-xmlish. 🤓

**Comment by wangrunji0408** [comment]
> I think the rfc is pretty descriptive. I think it's better than comment-based docs... But it's just my personal opinion.

From a user point of view, I think I need a quick start example. The RFC looks TL;DR. 🤣

File: src/planner/explain.rs [hunk 12]
```diff
@@ -65,210 +77,265 @@ impl<'a> Explain<'a> {
...
-                self.expr(lkeys),
-                self.expr(rkeys),
-                self.child(left),
-                self.child(right)
+            HashJoin([ty, lkeys, rkeys, left, right]) => {
+                let fields = btreemap! {
+                    "lhs" => self.expr(lkeys).pretty(),
+                    "rhs" => self.expr(rkeys).pretty(),
+                };
+                let eq = Pretty::childless_record("Equality", fields);
```
Comment (wangrunji0408) [inline src/planner/explain.rs:274]:
```suggestion
                let eq = Pretty::childless_record("=", fields);
```

File: tests/planner_test/tpch.planner.sql (hunk 8 above)
Comment (ice1000) [reply tests/planner_test/tpch.planner.sql:128]: How to handle BinExpr with enode = +, lhs = a scan node, rhs = an insert node? Or may I assume that such expressions will never appear?

What is the criteria for expressions to be valid for risinglight?

File: tests/planner_test/tpch.planner.sql (hunk 10 above)
Comment (wangrunji0408) [reply tests/planner_test/tpch.planner.sql:207]: https://github.com/risinglightdb/risinglight/pull/771/files#r1089681910

File: tests/planner_test/tpch.planner.sql (hunk 11 above)
Comment (wangrunji0408) [reply tests/planner_test/tpch.planner.sql:211]: Yeah. It might be the only way...

Comment (ice1000) [reply tests/planner_test/tpch.planner.sql:211]: Correct. I don't need any features of maps, and I only traverse them. I don't need the keys to be unique.

I should definitely use an associate list.

**Comment by ice1000** [comment]
> > I think the rfc is pretty descriptive. I think it's better than comment-based docs... But it's just my personal opinion.
> 
> From a user point of view, I think I need a quick start example. The RFC looks TL;DR. 🤣

I assume you're asking me to write a TL;DR for the rfc. I will do that tomorrow. You can stop me before that if that's not what you want.

File: tests/planner_test/tpch.planner.sql (hunk 8 above)
Comment (wangrunji0408) [reply tests/planner_test/tpch.planner.sql:128]: We should assume that will never happen, but I haven't find a proper way to ensure that in the egg framework.
In risinglight, expressions are e-node trees without any plan nodes. Though there's no formal definition yet. 😅

**Comment by ice1000** [comment]
@wangrunji0408 I have just came up with a perfect idea on addressing the binary expression problem. I'll do that tomorrow.

It requires extending the enum Pretty that makes the pretty printing result "binary" if two of its members are either binary or text.

In case they're not, or the line will become too long, the node will be displayed as a record with two fields.

This resolves the problem I mentioned just now.

File: tests/planner_test/tpch.planner.sql (hunk 8 above)
Comment (ice1000) [reply tests/planner_test/tpch.planner.sql:128]: I strongly believe that there should be a separate pretty printer for expressions. I know how to do it (and it's done properly in the Aya Prover), but it requires changing the AST structure.

**Comment by wangrunji0408** [comment]
> > > I think the rfc is pretty descriptive. I think it's better than comment-based docs... But it's just my personal opinion.
> > 
> > 
> > From a user point of view, I think I need a quick start example. The RFC looks TL;DR. 🤣
> 
> I assume you're asking me to write a TL;DR for the rfc. I will do that tomorrow. You can stop me before that if that's not what you want.

I mean the instructions for use, i.e. what code should I write to pretty print my struct. 🤔
A TL;DR for the design doc (RFC) is also nice to have.

**Comment by ice1000** [comment]
I see. That's totally reasonable. I'll see what I can do.

**Comment by ice1000** [comment]
I have partially fixed the problems of `pretty-xmlish`. It remains to deal with binary expressions... But IMO, in risingwave, there is no such design, right?

**Comment by wangrunji0408** [comment]
> I have partially fixed the problems of `pretty-xmlish`. It remains to deal with binary expressions... But IMO, in risingwave, there is no such design, right?

Yes. In RisingWave (and old version RisingLight), expressions and plan nodes are assembled into different enums separately.

**Comment by ice1000** [comment]
> > I have partially fixed the problems of `pretty-xmlish`. It remains to deal with binary expressions... But IMO, in risingwave, there is no such design, right?
> 
> Yes. In RisingWave (and old version RisingLight), expressions and plan nodes are assembled into different enums separately.

Why does risinglight migrated to this type unsafe approach then? 🥲

File: tests/planner_test/count.planner.sql [hunk 13]
```diff
@@ -2,19 +2,22 @@
...
-
-Projection { cost: 302.3, exprs: [ rowcount ] }              
-└── Aggregate { aggs: [ rowcount ], cost: 301, group_by: [] }
-    └── Scan { cost: 0, list: [], table: t }
+Projection { exprs: [ rowcount ], cost: 302.3 }
+└── Aggregate                                  
+    ├── aggs: [ rowcount ]                     
+    ├── group_by: []                           
+    ├── cost: 301                              
+    └── Scan { table: t, list: [], cost: 0 }
```
Comment (xxchan) [inline tests/planner_test/count.planner.sql:10]: Why is this changed? It seems you didn't change the line width and only reordered the fields. 🤔

**Review by xxchan — APPROVED** [review]
LGTM

**Comment by ice1000** [comment]
> A little question: why did you force-push? It made "Show changes since your last review" doesn't work 😇

Of course it's because I always forget to sign off my commits :cry:

File: tests/planner_test/count.planner.sql (hunk 13 above)
Comment (ice1000) [reply tests/planner_test/count.planner.sql:10]: Good question!

File: tests/planner_test/tpch.planner.sql (hunk 9 above)
Comment (ice1000) [reply tests/planner_test/tpch.planner.sql]: Fixed

**Comment by ice1000** [comment]
@wangrunji0408 please take a look. If you think it's okay to address binary expressions later, feel free to merge this.
If you want me to fix the problem and then merge it, let me know. I think we can first integrate this shit into risingwave which does not have this expression-statement mixed structure first :)

**Comment by wangrunji0408** [comment]
> Why does risinglight migrated to this type unsafe approach then? 🥲

It is an limitation of the egg framework. 🥲

File: tests/planner_test/tpch.planner.sql [hunk 14]
```diff
@@ -146,21 +192,29 @@ order by
...
-                Filter: (c_mktsegment = 'BUILDING') (cost=2700)
-                  Scan: customer[c_custkey, c_mktsegment] (cost=2000)
-              Filter: (1995-03-15 > o_orderdate) (cost=7500)
-                Scan: orders[o_orderkey, o_custkey, o_orderdate, o_shippriority] (cost=4000)
-          Projection: [l_orderkey, l_extendedprice, l_discount] (cost=10220)
-            Filter: (l_shipdate > 1995-03-15) (cost=7500)
-              Scan: lineitem[l_orderkey, l_extendedprice, l_discount, l_shipdate] (cost=4000)
+Projection { exprs: [ l_orderkey, sum, o_orderdate, o_shippriority ], cost: 63701.88 }                            
+└── TopN { limit: 10, offset: 0, order_by: [ desc, asc ], cost: 63647.88 }                                        
+    └── Aggregate { aggs: [ sum ], group_by: [ l_orderkey, o_orderdate, o_shippriority ], cost: 62224.105 }
```
Comment (wangrunji0408) [inline tests/planner_test/tpch.planner.sql:197]: The inner content of `sum` and `asc`/`desc` is still missing. 👀

**Comment by wangrunji0408** [comment]
Looks good to me. Please also fix the CI test before merging. Thx!

**Review by st1page — APPROVED** [review]
LGTM

**Comment by ice1000** [comment]
The doctest is failing!? 😲

File: tests/planner_test/tpch.planner.sql (hunk 14 above)
Comment (ice1000) [reply tests/planner_test/tpch.planner.sql]: Should be fixed now

**Review by wangrunji0408 — APPROVED** [review]
Success!

**Comment by ice1000** [comment]
Thank you @xxchan @wangrunji0408, w/o you guys I couldn't have discovered the (several) critical bug(s) of `pretty-xmlish`. I am also inspired to design some new features of it.

Thank you very much!
